lib/dao.js

```diff
@@ -598,11 +617,23 @@ DataAccessObject.findOrCreate = function findOrCreate(query, data, options, cb)
       }
     }
 
-    if (connector.findOrCreate.length === 5) {
-      connector.findOrCreate(modelName, query, self._forDB(data), options, findOrCreateCallback);
-    } else {
-      connector.findOrCreate(modelName, query, self._forDB(data), findOrCreateCallback);
-    }
+    data = removeUndefined(data);
+    var context = {
+      Model: Model,
+      data: data,
+      hookState: hookState,
+      options: options
```

Comment:
Please add a `where` property, I think the value should be `query.where`.

Notice that when the connector provides `connector.findOrCreate`, then we do create a new Model instance at https://github.com/strongloop/loopback-datasource-juggler/pull/586/files#diff-d717a01e40d8164976e4468af1bce663R666, thus `context.currentInstance` should be set too. Sorry for saying otherwise before.